src/network/subnet.cpp

```diff
@@ -0,0 +1,219 @@
+
+mp::Subnet parse(const std::string& cidr_string)
+try
+{
+    if (auto i = cidr_string.find('/'); i != std::string::npos)
+    {
+        mp::IPAddress addr{cidr_string.substr(0, i)};
+
+        auto cidr = std::stoul(cidr_string.substr(i + 1));
+        if (cidr >= 31)
+            throw std::invalid_argument(fmt::format(large_CIDR_err_fmt, cidr));
+
+        return mp::Subnet(addr, cidr);
+    }
+    throw std::invalid_argument(
+        fmt::format("CIDR address {} does not contain '/' seperator", cidr_string));
+}
+catch (const std::out_of_range& e)
+{
+    throw std::invalid_argument(fmt::format(large_CIDR_err_fmt, e.what()));
+}
+
+[[nodiscard]] mp::IPAddress get_subnet_mask(uint8_t cidr)
+{
+    using Octects = decltype(mp::IPAddress::octets);
+    static constexpr auto value_size = sizeof(Octects::value_type) * 8;
+
+    Octects octets{};
+    std::fill(octets.begin(), octets.end(), std::numeric_limits<Octects::value_type>::max());
+
+    if (cidr > value_size * octets.size())
+        throw std::out_of_range("CIDR too large for address space");
+
+    const uint8_t start_octet = cidr / value_size;
+    const uint8_t remain = cidr % value_size;
+
+    for (size_t i = start_octet; i < octets.size(); ++i)
+    {
+        octets[i] = 0;
+    }
+
+    if (remain != 0)
+    {
+        assert(start_octet < octets.size()); // sanity
+
+        // remain = 5, 1 << (8 - 5) = 00001000 -> 00000111 -> 11111000
+        octets[start_octet] = ~((1u << (value_size - remain)) - 1u);
+    }
+
+    return mp::IPAddress{octets};
+}
+
+[[nodiscard]] mp::IPAddress apply_mask(mp::IPAddress ip, uint8_t cidr)
+{
+    const auto mask = get_subnet_mask(cidr);
+    for (size_t i = 0; i < ip.octets.size(); ++i)
+    {
+        ip.octets[i] &= mask.octets[i];
+    }
+    return ip;
+}
+
+mp::Subnet generate_random_subnet(uint8_t cidr, mp::Subnet range)
+{
+    if (cidr >= 31)
+        throw std::invalid_argument(fmt::format(large_CIDR_err_fmt, cidr));
+
+    if (cidr < range.CIDR())
+        throw std::logic_error(fmt::format("A subnet with cidr {} cannot be contained by {}",
+                                           cidr,
+                                           range.as_string()));
+
+    // ex. 2^(24 - 16) = 256, [192.168.0.0/24, 192.168.255.0/24]
+    const size_t possibleSubnets = std::size_t{1} << (cidr - range.CIDR());
+
+    // narrowing conversion, possibleSubnets is guaranteed to be < 2^31 (4 bytes is safe)
+    static_assert(sizeof(decltype(MP_UTILS.random_int(0, possibleSubnets))) >= 4);
+
+    const auto subnet = static_cast<size_t>(MP_UTILS.random_int(0, possibleSubnets - 1));
+
+    // ex. 192.168.0.0 + (4 * 2^(32 - 24)) = 192.168.0.0 + 1024 = 192.168.4.0
+    mp::IPAddress id = range.identifier() + (subnet * (std::size_t{1} << (32 - cidr)));
+
+    return mp::Subnet{id, cidr};
+}
+} // namespace
+
+mp::Subnet::Subnet(IPAddress ip, uint8_t cidr) : id(apply_mask(ip, cidr)), cidr(cidr)
```

Comment:
What does `id` mean in this context? Network address? If so (the code looks that way), it's better to name it as such.